File: schemas/isn_notifications.py
"""ISN operational notification shapes — metadata-only, no raw safeguarding narrative."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

IsnNotificationSeverity = Literal["low", "medium", "high", "urgent"]
IsnNotificationType = Literal[
    "isn_safeguarding_alert",
    "isn_review_required",
    "isn_escalation_required",
    "isn_network_update",
    "isn_follow_up_due",
    "isn_professional_update",
    "isn_recording_linked_alert",
    "isn_manager_action_required",
]


class IsnNotificationItem(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    source: Literal["isn"] = "isn"
    type: IsnNotificationType | str
    title: str
    safe_summary: str
    severity: IsnNotificationSeverity = "medium"
    status: str = "open"
    route: str
    action_label: str | None = None
    child_id: int | None = None
    home_id: int | None = None
    created_at: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class IsnNotificationRoutes(BaseModel):
    model_config = ConfigDict(extra="ignore")

    safeguarding: str = "/safeguarding"
    alerts: str = "/safeguarding"
    recording_alerts: str = "/record/alerts"
    orb: str = "/assistant/orb?mode=safeguarding_themes"
    care_hub: str = "/command-centre"
    briefing: str = "/command-centre/briefing"


class IsnDigestTopItem(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    title: str
    safe_summary: str
    severity: IsnNotificationSeverity = "medium"
    type: str
    route: str
    action_label: str | None = None
    status: str = "open"


class IsnDigest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    generated_at: str
    available: bool = True
    total_open: int = 0
    urgent: int = 0
    high: int = 0
    review_required: int = 0
    escalation_required: int = 0
    follow_up_due: int = 0
    network_updates: int = 0
    linked_recording_alerts: int = 0
    top_items: list[IsnDigestTopItem] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)
    routes: IsnNotificationRoutes = Field(default_factory=IsnNotificationRoutes)
    privacy_notice: str = (
        "ISN summaries use metadata only. They support safeguarding oversight; "
        "they do not replace professional judgement or make referral decisions."
    )
    limitations: list[str] = Field(default_factory=list)


class IsnBadgeSummary(BaseModel):
    model_config = ConfigDict(extra="ignore")

    unread: int = 0
    urgent: int = 0
    review_required: int = 0
    available: bool = True


class IsnNotificationHealth(BaseModel):
    model_config = ConfigDict(extra="ignore")

    status: str = "ok"
    service: str = "isn_digest_service"
    persistence_available: bool = False
    storage_mode: str = "unknown"


IsnNotificationLifecycleStatus = Literal[
    "unread",
    "read",
    "acknowledged",
    "assigned",
    "resolved",
    "archived",
]

IsnNotificationActionType = Literal[
    "acknowledge",
    "assign",
    "resolve",
    "archive",
    "reopen",
    "create_intelligence_action",
]


class IsnNotificationActionRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    action: IsnNotificationActionType
    note: str | None = None
    owner_user_id: str | None = None
    owner_name: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class IsnNotificationActionResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    success: bool = True
    item_id: str
    action: str
    status: str = "acknowledged"
    message: str | None = None
    warning: str | None = None
    synced_to_os_state: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)

File: frontend-next/lib/os-api/isn-notifications.ts
export type IsnDigestTopItem = {
  id: string
  title: string
  safe_summary: string
  severity: 'low' | 'medium' | 'high' | 'urgent'
  type: string
  route: string
  action_label?: string | null
  status: string
}

export type IsnDigest = {
  generated_at: string
  available: boolean
  total_open: number
  urgent: number
  high: number
  review_required: number
  escalation_required: number
  follow_up_due: number
  network_updates: number
  linked_recording_alerts: number
  top_items: IsnDigestTopItem[]
  recommendations: string[]
  routes: {
    safeguarding: string
    alerts: string
    recording_alerts: string
    orb: string
    care_hub: string
    briefing: string
  }
  privacy_notice: string
  limitations: string[]
}

export type IsnBadgeSummary = {
  unread: number
  urgent: number
  review_required: number
  available: boolean
}

type ApiEnvelope<T> = { success?: boolean; data?: T }

const EMPTY_DIGEST: IsnDigest = {
  generated_at: '',
  available: false,
  total_open: 0,
  urgent: 0,
  high: 0,
  review_required: 0,
  escalation_required: 0,
  follow_up_due: 0,
  network_updates: 0,
  linked_recording_alerts: 0,
  top_items: [],
  recommendations: [],
  routes: {
    safeguarding: '/safeguarding',
    alerts: '/safeguarding',
    recording_alerts: '/record/alerts',
    orb: '/assistant/orb?mode=safeguarding_themes',
    care_hub: '/command-centre',
    briefing: '/command-centre/briefing'
  },
  privacy_notice: '',
  limitations: []
}

async function parseEnvelope<T>(response: Response, fallback: T): Promise<{ data: T; ok: boolean }> {
  if (!response.ok) return { data: fallback, ok: false }
  const payload = (await response.json().catch(() => ({}))) as ApiEnvelope<T> | T
  const envelope = payload as ApiEnvelope<T>
  return {
    data: envelope && typeof envelope === 'object' && 'data' in envelope ? (envelope.data as T) : (payload as T),
    ok: true
  }
}

export async function getIsnNotificationHealth() {
  const response = await fetch('/api/isn/notifications/health', {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope(response, { status: 'unavailable', service: 'isn_digest_service' })
}

export async function getIsnDigest() {
  const response = await fetch('/api/isn/notifications/digest', {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope(response, EMPTY_DIGEST)
}

export async function getIsnBadgeSummary() {
  const response = await fetch('/api/isn/notifications/badge-summary', {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope(response, { unread: 0, urgent: 0, review_required: 0, available: false })
}

export async function listIsnNotificationItems(limit = 30) {
  const response = await fetch(`/api/isn/notifications/items?limit=${limit}`, {
    credentials: 'include',
    cache: 'no-store'
  })
  return parseEnvelope(response, [] as IsnDigestTopItem[])
}

export function isnOrbHref(mode = 'safeguarding_themes', query?: string) {
  const params = new URLSearchParams({ mode })
  if (query) params.set('q', query)
  return `/assistant/orb?${params.toString()}`
}
